pkg/cluster/cluster.go

```diff
@@ -14,3 +22,51 @@ type Cluster struct {
 	RESTClient *rest.RESTClient
 	Cache      cache.Cache
 }
+
+// LookupMember takes in a list of member clusters and a Request, it then searches for the member cluster that triggered the request.
+// returns memberCluster/true/nil in case the Object was found on one of the given member clusters
+// returns nil/false/nil in case the Object was NOT found on any of the given member clusters
+// returns nil/false/error in case the Object was NOT found on any of the given member clusters and there was an error while trying to read one or more of the memberclusters
+//
+// CAVEAT: if the object was not found and there is an error reading from more than one member cluster, only the last read error will be returned.
+func LookupMember(MemberClusters map[string]Cluster, request ctrl.Request, object runtimeclient.Object) (Cluster, bool, error) {
+	var memberClusterWithObject Cluster
+	var err, getError error
+	for _, memberCluster := range MemberClusters {
+		err = memberCluster.Client.Get(context.TODO(), types.NamespacedName{
+			Namespace: request.Namespace,
+			Name:      request.Name,
+		}, object)
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				// Error reading the object
+				getError = err // save the error and return it only later in case the Object was not found on any of the clusters
+			}
+
+			//  Object not found on current member cluster
+			continue
+		}
+
+		// save the member cluster on which the Object was found
+		memberClusterWithObject = memberCluster
+		break // exit once found
+	}
+
+	// if we haven't found the Object, the memberCluster is an empty struct,
+	// we can't do the same check on the runtimeclient.Object since for some reason it has the TypeMeta field on it
+	// and the check fails.
+	if reflect.ValueOf(memberClusterWithObject).IsZero() {
+		if getError != nil {
+			// if we haven't found the Object, and we got an error while trying to retrieve it
+			// then let's return the error
+			return memberClusterWithObject, false, errs.Wrap(getError, fmt.Sprintf("unable to get the current %T", object))
+		} else if err != nil && errors.IsNotFound(err) {
+			// if we exited with a notFound error
+			// it means that we couldn't find the Object on any of the given member clusters
+			// let's just return false (as not found) and no error
+			return memberClusterWithObject, false, nil
+		}
+	}
+	// Object found, return the member cluster that has it
+	return memberClusterWithObject, true, nil
+}
```

Comment:
I tried to simplify the code and also fixed some minor issues. Take a look:
```suggestion
func LookupMember(memberClusters map[string]Cluster, request ctrl.Request, object runtimeclient.Object) (Cluster, bool, error) {
	var memberClusterWithObject Cluster
	var getError error
	var found bool
	for _, memberCluster := range memberClusters {
		err := memberCluster.Client.Get(context.TODO(), types.NamespacedName{
			Namespace: request.Namespace,
			Name:      request.Name,
		}, object)
		if err != nil {
			if !errors.IsNotFound(err) {
				// Error reading the object
				getError = err // save the error and return it only later in case the Object was not found on any of the clusters
			}

			//  Object not found on current member cluster
			continue
		}

		// save the member cluster on which the Object was found
		memberClusterWithObject = memberCluster
		found = true
		break // exit once found
	}

	if !found {
		if getError != nil {
			// if we haven't found the Object, and we got an error while trying to retrieve it
			// then let's return the error
			return memberClusterWithObject, false, errs.Wrap(getError, fmt.Sprintf("unable to get the current %T", object))
		}
		// Not found. No Error.
		return nil, false, nil
	}
	// Object found, return the member cluster that has it. Also if we got an error when trying to connec to at least one of them member then return it too.
	return memberClusterWithObject, true, getError
}
```

Reply:
l like having the variable `found` to observe if the cluster was found, and use that variable in the return statement. It makes it easier to read as well as removes hardcoding the return values.

Reply:
OK. I updated the suggestion.

Reply:
At first I thought about the same, just didn't want to introduce a new variable (`found`), but if it makes the code easier to read let's keep it. 

If we want to return `nil` instead of `memberClusterWithObject` I would need to return a pointer and change the signature of all the functions that gets it as input to accept a pointer, I suggest we do this in a separate PR. 

plz check https://github.com/codeready-toolchain/host-operator/pull/855/commits/e618eac3a659e96ac6d7863fbba58feb743611ee

Thanks for all the suggestions!

Reply:
OK. You can keep the memberSluster object as is in the signature. Doesn't have to be a pointer. However please fix the descritpion then. It says it will return nil if not found:
> returns nil/false/nil in case the Object was NOT found on any of the given member clusters
> returns nil/false/error in case the Object was NOT found on any of the given member clusters...

Reply:
Ok thanks for pointing that out, it returns an empty struct and not a nil indeed. I've changed it here: https://github.com/codeready-toolchain/host-operator/pull/855/commits/09cb5e87138939915d64006d0e9a11de6184a206 I guess this is what you meant.